In [54]:
import pandas as pd
ndf = pd.read_csv('nhappy_0814.csv')

ndf['country']= ndf['country'].astype("category")
ndf['year']= ndf['year'].astype("category")
ndf['country_num']= ndf['country_num'].astype("category")
ndf['continent']= ndf['continent'].astype("category")

from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
x= ndf[['log_gdp', 'social', 'life_exp', 'freedom','generosity', 'corruption', 
       'schooling', 'alcohol', 'suicide','obesity', 'birth', 'mental']]
xnor = MinMaxScaler().fit_transform(x)
dfnor = pd.DataFrame(xnor)
dfnor.columns = ['log_gdp', 'social', 'life_exp', 'freedom', 'generosity', 'corruption', 'schooling', 'alcohol',
                     'suicide','obesity', 'birth', 'mental']

# 정규화 안한 컬럼 dfnor에 추가
dfnor[['HS','year','country','country_num','continent']]= ndf[['HS','year','country','country_num','continent']]
dfnor

log_gdp    social  life_exp   freedom  generosity  corruption  \
0     0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   
1     0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   
2     0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   
3     0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   
4     0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   
...        ...       ...       ...       ...         ...         ...   
1876  0.559944  0.880918  0.750446  0.760660    0.253630    0.856540   
1877  0.379613  0.474892  0.587054  0.757909    0.299129    0.840717   
1878  0.853980  0.949785  0.888393  0.958735    0.474347    0.154008   
1879  0.505087  0.846485  0.616071  0.792297    0.583737    0.648734   
1880  0.514662  0.771879  0.533482  0.759285    0.140368    0.839662   

      schooling   alcohol   suicide   obesity     birth    mental     HS  \
0      0.421875  0.269256  0.215085  0.507264  0.206485  0.530582  5.819   
1      0.429688  0.075482  0.093262  0.184019  0.479963  0.179361  5.057   
2      0.117188  0.087591  0.065017  0.071025  0.691649  0.086241  3.330   
3      0.750000  0.071034  0.525613  0.343019  0.176201  0.493112  4.331   
4      0.304688  0.000365  0.134452  0.135190  0.163047  0.419619  4.633   
...         ...       ...       ...       ...       ...       ...    ...   
1876   0.484375  0.002508  0.169182  0.352300  0.138269  0.434382  6.464   
1877   0.375000  0.078060  0.306743  0.143261  0.195626  0.318333  4.424   
1878   0.882812  0.319349  0.164167  0.270379  0.106455  0.515135  7.594   
1879   0.140625  0.056216  0.145703  0.211057  0.154940  0.447314  4.939   
1880   0.429688  0.085750  0.176421  0.217111  0.366014  0.244110  4.441   

      year     country country_num continent  
0     2018   Nicaragua         558     Nor.A  
1     2012       Ghana         288        AF  
2     2006       Benin         204        AF  
3     2017   Sri Lanka         144        AS  
4     2015  Bangladesh          50        AS  
...    ...         ...         ...       ...  
1876  2011    Colombia         170        SA  
1877  2014       India         356        AS  
1878  2017     Denmark         208        EU  
1879  2014      Bhutan          64        AS  
1880  2017     Namibia         516        AF  

[1881 rows x 17 columns]

In [ ]:
# 나라 원핫인코딩 

In [55]:
df_country = pd.get_dummies(dfnor['country'])
df_country

Afghanistan  Albania  Algeria  Angola  Argentina  Armenia  Australia  \
0               0        0        0       0          0        0          0   
1               0        0        0       0          0        0          0   
2               0        0        0       0          0        0          0   
3               0        0        0       0          0        0          0   
4               0        0        0       0          0        0          0   
...           ...      ...      ...     ...        ...      ...        ...   
1876            0        0        0       0          0        0          0   
1877            0        0        0       0          0        0          0   
1878            0        0        0       0          0        0          0   
1879            0        0        0       0          0        0          0   
1880            0        0        0       0          0        0          0   

      Austria  Azerbaijan  Bahrain  ...  United Arab Emirates  United Kingdom  \
0           0           0        0  ...                     0               0   
1           0           0        0  ...                     0               0   
2           0           0        0  ...                     0               0   
3           0           0        0  ...                     0               0   
4           0           0        0  ...                     0               0   
...       ...         ...      ...  ...                   ...             ...   
1876        0           0        0  ...                     0               0   
1877        0           0        0  ...                     0               0   
1878        0           0        0  ...                     0               0   
1879        0           0        0  ...                     0               0   
1880        0           0        0  ...                     0               0   

      United States  Uruguay  Uzbekistan  Venezuela  Vietnam  Yemen  Zambia  \
0                 0        0           0          0        0      0       0   
1                 0        0           0          0        0      0       0   
2                 0        0           0          0        0      0       0   
3                 0        0           0          0        0      0       0   
4                 0        0           0          0        0      0       0   
...             ...      ...         ...        ...      ...    ...     ...   
1876              0        0           0          0        0      0       0   
1877              0        0           0          0        0      0       0   
1878              0        0           0          0        0      0       0   
1879              0        0           0          0        0      0       0   
1880              0        0           0          0        0      0       0   

      Zimbabwe  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
1876         0  
1877         0  
1878         0  
1879         0  
1880         0  

[1881 rows x 153 columns]

In [56]:
dfnor = pd.concat([dfnor, df_country], axis=1)
dfnor

log_gdp    social  life_exp   freedom  generosity  corruption  \
0     0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   
1     0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   
2     0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   
3     0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   
4     0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   
...        ...       ...       ...       ...         ...         ...   
1876  0.559944  0.880918  0.750446  0.760660    0.253630    0.856540   
1877  0.379613  0.474892  0.587054  0.757909    0.299129    0.840717   
1878  0.853980  0.949785  0.888393  0.958735    0.474347    0.154008   
1879  0.505087  0.846485  0.616071  0.792297    0.583737    0.648734   
1880  0.514662  0.771879  0.533482  0.759285    0.140368    0.839662   

      schooling   alcohol   suicide   obesity  ...  United Arab Emirates  \
0      0.421875  0.269256  0.215085  0.507264  ...                     0   
1      0.429688  0.075482  0.093262  0.184019  ...                     0   
2      0.117188  0.087591  0.065017  0.071025  ...                     0   
3      0.750000  0.071034  0.525613  0.343019  ...                     0   
4      0.304688  0.000365  0.134452  0.135190  ...                     0   
...         ...       ...       ...       ...  ...                   ...   
1876   0.484375  0.002508  0.169182  0.352300  ...                     0   
1877   0.375000  0.078060  0.306743  0.143261  ...                     0   
1878   0.882812  0.319349  0.164167  0.270379  ...                     0   
1879   0.140625  0.056216  0.145703  0.211057  ...                     0   
1880   0.429688  0.085750  0.176421  0.217111  ...                     0   

      United Kingdom  United States Uruguay Uzbekistan Venezuela Vietnam  \
0                  0              0       0          0         0       0   
1                  0              0       0          0         0       0   
2                  0              0       0          0         0       0   
3                  0              0       0          0         0       0   
4                  0              0       0          0         0       0   
...              ...            ...     ...        ...       ...     ...   
1876               0              0       0          0         0       0   
1877               0              0       0          0         0       0   
1878               0              0       0          0         0       0   
1879               0              0       0          0         0       0   
1880               0              0       0          0         0       0   

      Yemen  Zambia  Zimbabwe  
0         0       0         0  
1         0       0         0  
2         0       0         0  
3         0       0         0  
4         0       0         0  
...     ...     ...       ...  
1876      0       0         0  
1877      0       0         0  
1878      0       0         0  
1879      0       0         0  
1880      0       0         0  

[1881 rows x 170 columns]

In [57]:
dfnor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1881 entries, 0 to 1880
Columns: 170 entries, log_gdp to Zimbabwe
dtypes: category(4), float64(13), uint8(153)
memory usage: 494.6 KB


In [58]:
import pandas as pd
import seaborn as sns
from statsmodels.formula.api import ols
import numpy as np
from scipy import stats
from scipy.stats import norm
import matplotlib.pyplot as plt

# model 1 : country 포함

In [39]:
model1 = ols("np.square(HS)~ country+ log_gdp + social+ life_exp + freedom+ generosity+ corruption+ schooling+ alcohol+ suicide+ obesity+ birth+ mental", data=dfnor)
res1 = model1.fit()
res1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.square(HS)   R-squared:                       0.923
Model:                            OLS   Adj. R-squared:                  0.916
Method:                 Least Squares   F-statistic:                     125.3
Date:                Tue, 17 Aug 2021   Prob (F-statistic):               0.00
Time:                        17:27:54   Log-Likelihood:                -5003.4
No. Observations:                1881   AIC:                         1.034e+04
Df Residuals:                    1716   BIC:                         1.125e+04
Df Model:                         164                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               3.8209      4.002      0.955      0.340      -4.029      11.671
country[T.Albania]                     -0.2323      2.648     -0.088      0.930      -5.426       4.961
country[T.Algeria]                      3.6627      2.482      1.475      0.140      -1.206       8.531
country[T.Angola]                     -14.1848      2.763     -5.134      0.000     -19.604      -8.766
country[T.Argentina]                    2.1454      2.953      0.726      0.468      -3.647       7.938
country[T.Armenia]                     -1.5153      2.855     -0.531      0.596      -7.114       4.084
country[T.Australia]                    3.7333      4.041      0.924      0.356      -4.193      11.660
country[T.Austria]                      2.9948      3.785      0.791      0.429      -4.429      10.419
country[T.Azerbaijan]                  -2.3324      2.790     -0.836      0.403      -7.806       3.141
country[T.Bahrain]                     -7.7592      3.830     -2.026      0.043     -15.271      -0.247
country[T.Bangladesh]                   3.5253      2.420      1.456      0.145      -1.222       8.273
country[T.Belarus]                      4.9756      4.170      1.193      0.233      -3.204      13.155
country[T.Belgium]                     -0.8825      3.918     -0.225      0.822      -8.566       6.801
country[T.Belize]                      14.9750      3.491      4.290      0.000       8.128      21.822
country[T.Benin]                       -4.6452      1.737     -2.674      0.008      -8.053      -1.238
country[T.Bhutan]                      -9.8863      3.433     -2.880      0.004     -16.619      -3.154
country[T.Bolivia]                      5.8552      2.276      2.573      0.010       1.391      10.319
country[T.Bosnia and Herzegovina]       2.3909      2.830      0.845      0.398      -3.160       7.942
country[T.Botswana]                   -22.7712      2.973     -7.658      0.000     -28.603     -16.939
country[T.Brazil]                      10.9534      2.736      4.003      0.000       5.587      16.320
country[T.Bulgaria]                    -9.2491      3.271     -2.828      0.005     -15.664      -2.834
country[T.Burkina Faso]                -3.1190      1.754     -1.778      0.076      -6.559       0.321
country[T.Burundi]                      7.8382      2.293      3.418      0.001       3.341      12.336
country[T.Cambodia]                    -4.8470      2.311     -2.097      0.036      -9.380      -0.314
country[T.Cameroon]                    -2.2065      1.871     -1.179      0.238      -5.876       1.463
country[T.Canada]                       5.7194      3.829      1.494      0.135      -1.790      13.229
country[T.Central African Republic]     3.5628      2.243      1.58

## X, y 데이터 분리

In [59]:
X_data = dfnor.drop(['HS','year','country','country_num','continent'], axis = 1)
y = dfnor[['HS']]

X_data.head()

log_gdp    social  life_exp   freedom  generosity  corruption  schooling  \
0  0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   0.421875   
1  0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   0.429688   
2  0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   0.117188   
3  0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   0.750000   
4  0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   0.304688   

    alcohol   suicide   obesity  ...  United Arab Emirates  United Kingdom  \
0  0.269256  0.215085  0.507264  ...                     0               0   
1  0.075482  0.093262  0.184019  ...                     0               0   
2  0.087591  0.065017  0.071025  ...                     0               0   
3  0.071034  0.525613  0.343019  ...                     0               0   
4  0.000365  0.134452  0.135190  ...                     0               0   

   United States  Uruguay  Uzbekistan  Venezuela  Vietnam  Yemen  Zambia  \
0              0        0           0          0        0      0       0   
1              0        0           0          0        0      0       0   
2              0        0           0          0        0      0       0   
3              0        0           0          0        0      0       0   
4              0        0           0          0        0      0       0   

   Zimbabwe  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 165 columns]

## train, test 데이터 분리

In [60]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X_data, y_target,
                                                  test_size=0.3, random_state=4)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(1316, 165)
(1316, 1)
(565, 165)
(565, 1)


## 1. Linear Regression

In [42]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_train = lr.predict(X_train)
y_pred_test = lr.predict(X_test)

train_mse = mean_squared_error(y_train, y_pred_train)
train_rmse = np.sqrt(train_mse)

test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

print('lr.coef_: {}'.format(lr.coef_))
print('Train MSE:{0:.3f}, RMSE:{1:3.3F}'.format(train_mse, train_rmse))
print('Train Variance score: {0:.3f}'.format(r2_score(y_train, y_pred_train)))
print('Test MSE:{0:.3f}, RMSE:{1:3.3F}'.format(test_mse, test_rmse))
print('Test Variance score: {0:.3f}'.format(r2_score(y_test, y_pred_test)))

lr.coef_: [[ 5.68866818e+00  1.24817854e+00 -6.67942988e-01  6.37224601e-01
   3.15810159e-01 -3.04588380e-01 -2.10787058e-01 -8.42809113e-01
   1.19915591e+00 -1.34881594e+00  1.23242635e+00 -3.59076995e-01
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11
   2.87768801e+11  2.87768801e+11  2.87768801e+11  2.87768801e+11


## 2. Ridge Regression

In [61]:
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score
ridge = Ridge(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.001, 0.005, 0.01, 0.015, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(ridge, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.1}
-0.14025719008940585
0.37450926569232684
0.39111393469538003


## 3. Lasso Regression

In [62]:
from sklearn.linear_model import Lasso
lasso = Lasso(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.0001, 0.005, 0.01, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(lasso, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.0001}
-0.1395883018875552
0.3736151788773513
0.39163181020107013


## 4. ElasticNet Regression

In [63]:
from sklearn.linear_model import ElasticNet

elasticnet = ElasticNet(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.0001, 0.005, 0.01, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(elasticnet, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.0001}
-0.13979725436211482
0.37389471026228066
0.391385079918207


## MODEL 2 : CONTINENT 포함

In [65]:
import pandas as pd
ndf = pd.read_csv('nhappy_0814.csv')

ndf['country']= ndf['country'].astype("category")
ndf['year']= ndf['year'].astype("category")
ndf['country_num']= ndf['country_num'].astype("category")
ndf['continent']= ndf['continent'].astype("category")

from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
x= ndf[['log_gdp', 'social', 'life_exp', 'freedom','generosity', 'corruption', 
       'schooling', 'alcohol', 'suicide','obesity', 'birth', 'mental']]
xnor = MinMaxScaler().fit_transform(x)
dfnor = pd.DataFrame(xnor)
dfnor.columns = ['log_gdp', 'social', 'life_exp', 'freedom', 'generosity', 'corruption', 'schooling', 'alcohol',
                     'suicide','obesity', 'birth', 'mental']

# 정규화 안한 컬럼 dfnor에 추가
dfnor[['HS','year','country','country_num','continent']]= ndf[['HS','year','country','country_num','continent']]
dfnor

log_gdp    social  life_exp   freedom  generosity  corruption  \
0     0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   
1     0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   
2     0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   
3     0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   
4     0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   
...        ...       ...       ...       ...         ...         ...   
1876  0.559944  0.880918  0.750446  0.760660    0.253630    0.856540   
1877  0.379613  0.474892  0.587054  0.757909    0.299129    0.840717   
1878  0.853980  0.949785  0.888393  0.958735    0.474347    0.154008   
1879  0.505087  0.846485  0.616071  0.792297    0.583737    0.648734   
1880  0.514662  0.771879  0.533482  0.759285    0.140368    0.839662   

      schooling   alcohol   suicide   obesity     birth    mental     HS  \
0      0.421875  0.269256  0.215085  0.507264  0.206485  0.530582  5.819   
1      0.429688  0.075482  0.093262  0.184019  0.479963  0.179361  5.057   
2      0.117188  0.087591  0.065017  0.071025  0.691649  0.086241  3.330   
3      0.750000  0.071034  0.525613  0.343019  0.176201  0.493112  4.331   
4      0.304688  0.000365  0.134452  0.135190  0.163047  0.419619  4.633   
...         ...       ...       ...       ...       ...       ...    ...   
1876   0.484375  0.002508  0.169182  0.352300  0.138269  0.434382  6.464   
1877   0.375000  0.078060  0.306743  0.143261  0.195626  0.318333  4.424   
1878   0.882812  0.319349  0.164167  0.270379  0.106455  0.515135  7.594   
1879   0.140625  0.056216  0.145703  0.211057  0.154940  0.447314  4.939   
1880   0.429688  0.085750  0.176421  0.217111  0.366014  0.244110  4.441   

      year     country country_num continent  
0     2018   Nicaragua         558     Nor.A  
1     2012       Ghana         288        AF  
2     2006       Benin         204        AF  
3     2017   Sri Lanka         144        AS  
4     2015  Bangladesh          50        AS  
...    ...         ...         ...       ...  
1876  2011    Colombia         170        SA  
1877  2014       India         356        AS  
1878  2017     Denmark         208        EU  
1879  2014      Bhutan          64        AS  
1880  2017     Namibia         516        AF  

[1881 rows x 17 columns]

In [77]:
model2 = ols("np.square(HS)~ continent+ log_gdp + social+ life_exp + freedom+ generosity+ corruption+ schooling+ alcohol+ suicide+ obesity+ birth+ mental", data=dfnor)
res2 = model2.fit()
res2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.square(HS)   R-squared:                       0.814
Model:                            OLS   Adj. R-squared:                  0.812
Method:                 Least Squares   F-statistic:                     478.3
Date:                Tue, 17 Aug 2021   Prob (F-statistic):               0.00
Time:                        17:41:16   Log-Likelihood:                -5833.9
No. Observations:                1881   AIC:                         1.170e+04
Df Residuals:                    1863   BIC:                         1.180e+04
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             -9.5221      1.891     -5.036      0.000     -13.230      -5.814
continent[T.AS]        1.3711      0.496      2.763      0.006       0.398       2.344
continent[T.EU]        4.3099      0.678      6.352      0.000       2.979       5.641
continent[T.Nor.A]     9.0211      0.665     13.558      0.000       7.716      10.326
continent[T.OC]        6.3002      1.239      5.085      0.000       3.870       8.730
continent[T.SA]        7.7305      0.662     11.685      0.000       6.433       9.028
log_gdp               30.5554      1.537     19.884      0.000      27.542      33.569
social                10.7519      1.150      9.346      0.000       8.496      13.008
life_exp              11.0239      2.198      5.014      0.000       6.712      15.336
freedom                6.9382      0.858      8.089      0.000       5.256       8.620
generosity             7.5365      0.974      7.740      0.000       5.627       9.446
corruption            -7.1085      0.849     -8.377      0.000      -8.773      -5.444
schooling              1.7851      1.061      1.683      0.093      -0.295       3.865
alcohol                4.6450      1.175      3.952      0.000       2.340       6.950
suicide               -4.4590      1.392     -3.203      0.001      -7.189      -1.729
obesity               -7.8470      0.986     -7.961      0.000      -9.780      -5.914
birth                 13.5701      1.436      9.450      0.000      10.754      16.386
mental                 5.4303      1.538      3.530      0.000       2.413       8.447
==============================================================================
Omnibus:                        5.966   Durbin-Watson:                   2.060
Prob(Omnibus):                  0.051   Jarque-Bera (JB):                7.233
Skew:                          -0.003   Prob(JB):                       0.0269
Kurtosis:                       3.304   Cond. No.                         49.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [66]:
# continent 원핫인코딩
df_continent = pd.get_dummies(dfnor['continent'])
df_continent

AF  AS  EU  Nor.A  OC  SA
0      0   0   0      1   0   0
1      1   0   0      0   0   0
2      1   0   0      0   0   0
3      0   1   0      0   0   0
4      0   1   0      0   0   0
...   ..  ..  ..    ...  ..  ..
1876   0   0   0      0   0   1
1877   0   1   0      0   0   0
1878   0   0   1      0   0   0
1879   0   1   0      0   0   0
1880   1   0   0      0   0   0

[1881 rows x 6 columns]

In [67]:
dfnor = pd.concat([dfnor, df_continent], axis=1)
dfnor

log_gdp    social  life_exp   freedom  generosity  corruption  \
0     0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   
1     0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   
2     0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   
3     0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   
4     0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   
...        ...       ...       ...       ...         ...         ...   
1876  0.559944  0.880918  0.750446  0.760660    0.253630    0.856540   
1877  0.379613  0.474892  0.587054  0.757909    0.299129    0.840717   
1878  0.853980  0.949785  0.888393  0.958735    0.474347    0.154008   
1879  0.505087  0.846485  0.616071  0.792297    0.583737    0.648734   
1880  0.514662  0.771879  0.533482  0.759285    0.140368    0.839662   

      schooling   alcohol   suicide   obesity  ...  year     country  \
0      0.421875  0.269256  0.215085  0.507264  ...  2018   Nicaragua   
1      0.429688  0.075482  0.093262  0.184019  ...  2012       Ghana   
2      0.117188  0.087591  0.065017  0.071025  ...  2006       Benin   
3      0.750000  0.071034  0.525613  0.343019  ...  2017   Sri Lanka   
4      0.304688  0.000365  0.134452  0.135190  ...  2015  Bangladesh   
...         ...       ...       ...       ...  ...   ...         ...   
1876   0.484375  0.002508  0.169182  0.352300  ...  2011    Colombia   
1877   0.375000  0.078060  0.306743  0.143261  ...  2014       India   
1878   0.882812  0.319349  0.164167  0.270379  ...  2017     Denmark   
1879   0.140625  0.056216  0.145703  0.211057  ...  2014      Bhutan   
1880   0.429688  0.085750  0.176421  0.217111  ...  2017     Namibia   

      country_num continent AF AS EU  Nor.A  OC  SA  
0             558     Nor.A  0  0  0      1   0   0  
1             288        AF  1  0  0      0   0   0  
2             204        AF  1  0  0      0   0   0  
3             144        AS  0  1  0      0   0   0  
4              50        AS  0  1  0      0   0   0  
...           ...       ... .. .. ..    ...  ..  ..  
1876          170        SA  0  0  0      0   0   1  
1877          356        AS  0  1  0      0   0   0  
1878          208        EU  0  0  1      0   0   0  
1879           64        AS  0  1  0      0   0   0  
1880          516        AF  1  0  0      0   0   0  

[1881 rows x 23 columns]

## X, y 데이터 분리

In [71]:
X_data = dfnor.drop(['HS','year','country','country_num','continent'], axis = 1)
y = dfnor[['HS']]

X_data.head()

log_gdp    social  life_exp   freedom  generosity  corruption  schooling  \
0  0.401356  0.809182  0.785714  0.741403    0.333011    0.715190   0.421875   
1  0.351486  0.566714  0.503571  0.579092    0.363020    0.910338   0.429688   
2  0.245562  0.222382  0.397321  0.442916    0.313650    0.796414   0.117188   
3  0.559545  0.764706  0.774554  0.782669    0.415295    0.853376   0.750000   
4  0.315380  0.446198  0.680804  0.766162    0.258470    0.723629   0.304688   

    alcohol   suicide   obesity     birth    mental  AF  AS  EU  Nor.A  OC  SA  
0  0.269256  0.215085  0.507264  0.206485  0.530582   0   0   0      1   0   0  
1  0.075482  0.093262  0.184019  0.479963  0.179361   1   0   0      0   0   0  
2  0.087591  0.065017  0.071025  0.691649  0.086241   1   0   0      0   0   0  
3  0.071034  0.525613  0.343019  0.176201  0.493112   0   1   0      0   0   0  
4  0.000365  0.134452  0.135190  0.163047  0.419619   0   1   0      0   0   0

## train, test 데이터 분리

In [72]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X_data, y_target,
                                                  test_size=0.3, random_state=4)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(1316, 18)
(1316, 1)
(565, 18)
(565, 1)


## 1. Linear Regression

In [73]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_train = lr.predict(X_train)
y_pred_test = lr.predict(X_test)

train_mse = mean_squared_error(y_train, y_pred_train)
train_rmse = np.sqrt(train_mse)

test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

print('lr.coef_: {}'.format(lr.coef_))
print('Train MSE:{0:.3f}, RMSE:{1:3.3F}'.format(train_mse, train_rmse))
print('Train Variance score: {0:.3f}'.format(r2_score(y_train, y_pred_train)))
print('Test MSE:{0:.3f}, RMSE:{1:3.3F}'.format(test_mse, test_rmse))
print('Test Variance score: {0:.3f}'.format(r2_score(y_test, y_pred_test)))

lr.coef_: [[ 2.63252215  1.24957338  1.14717676  0.66744971  0.54333486 -0.42525415
  -0.04228437  0.44888863 -0.2941654  -0.53353524  1.1352664   0.46008851
  -0.46948225 -0.27537408 -0.05057957  0.35434289  0.10143165  0.33966137]]
Train MSE:0.243, RMSE:0.493
Train Variance score: 0.808
Test MSE:0.259, RMSE:0.509
Test Variance score: 0.792


## 2. Ridge Regression

In [74]:
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score
ridge = Ridge(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.001, 0.005, 0.01, 0.015, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(ridge, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.1}
-0.25002723561305007
0.5000272348713118
0.508879082220966


## 3. Lasso Regression

In [75]:
from sklearn.linear_model import Lasso
lasso = Lasso(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.0001, 0.005, 0.01, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(lasso, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.0001}
-0.25005334962366
0.5000533467777812
0.5087843229715936


## 4. ElasticNet Regression

In [76]:
from sklearn.linear_model import ElasticNet

elasticnet = ElasticNet(random_state=4)

cv_method = KFold(n_splits=10)
params = {
    'alpha': [0.0001, 0.005, 0.01, 0.1, 1, 10]
}
dt_grid_cv = GridSearchCV(elasticnet, param_grid=params, scoring='neg_mean_squared_error', cv=cv_method, n_jobs=-1)
dt_grid_cv.fit(X_train, y_train)

y_pred_test = dt_grid_cv.predict(X_test)
test_mse = mean_squared_error(y_test, y_pred_test)
test_rmse = np.sqrt(test_mse)

neg_mse = dt_grid_cv.best_score_
mse = -1 * neg_mse
rmse = np.sqrt(mse)

print(dt_grid_cv.best_params_)
print(dt_grid_cv.best_score_)
print(rmse)
print(test_rmse)

{'alpha': 0.0001}
-0.2500340067295821
0.5000340055732031
0.5088393106728601


# 결론 

## model1 (country 포함) 
#### - R^2 : 0.923
#### - test 데이터 RMSE : 약 0.39 


## model2 (continent 포함) 
#### - R^2 : 0.814
#### - test 데이터 RMSE : 약 0.51